provider/openstack/firewaller.go

```diff
@@ -70,16 +71,342 @@ type firewallerFactory struct {
 
 // GetFirewaller implements FirewallerFactory
 func (f *firewallerFactory) GetFirewaller(env environs.Environ) Firewaller {
-	return &defaultFirewaller{environ: env.(*Environ)}
+	return &switchingFirewaller{env: env.(*Environ)}
+}
+
+type switchingFirewaller struct {
```

Comment:
We need to avoid making API calls in EnvironProvider.Open, which would otherwise be the natural place to do a single auth call to get the endpoints.

We could alternatively defer the firewaller construction to each firewaller method call in the Environ. The area of firewalling needs shaking up in general. I would like to split it off like storage providers, and just have Environ be a source of firewall provider(s). Then the firewaller worker would construct them, and that bit could be dynamic.

I'm going to leave it for now, because this is meant to be a temporary thing anyway. If it turns out not to be, we should find a way to tidy it up. If we can drop the Nova networking bit soon, then the switching impls can go away.